smartapps/opent2t/opent2t-smartapp-test.src/opent2t-smartapp-test.groovy

```diff
@@ -0,0 +1,355 @@
+/**
+ *  OpenT2T SmartApp Test
+ *
+ *  Copyright 2016 OpenT2T
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License. You may obtain a copy of the License at:
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed
+ *  on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License
+ *  for the specific language governing permissions and limitations under the License.
+ *
+ */
+definition(
+    name: "OpenT2T SmartApp Test",
+    namespace: "opent2t",
+    author: "OpenT2T",
+    description: "Test app to test end to end SmartThings scenarios via OpenT2T",
+    category: "SmartThings Labs",
+    iconUrl: "https://s3.amazonaws.com/smartapp-icons/Convenience/Cat-Convenience.png",
+    iconX2Url: "https://s3.amazonaws.com/smartapp-icons/Convenience/Cat-Convenience@2x.png",
+    iconX3Url: "https://s3.amazonaws.com/smartapp-icons/Convenience/Cat-Convenience@2x.png")
+
+//Device Inputs
+preferences {
+section("Allow OpenT2T Test SmartApp to control these things...") {
+		input "switches", "capability.switch", title: "Which Switches and Lights?", multiple: true, required: false
+		input "motionSensors", "capability.motionSensor", title: "Which Motion Sensors?", multiple: true, required: false
+		input "presenceSensors", "capability.presenceSensor", title: "Which Presence Sensors", multiple: true, required: false
+        input "contactSensors", "capability.contactSensor", title: "Which Contact Sensors", multiple: true, required: false
+        input "waterSensors", "capability.waterSensor", title: "Which Water Leak Sensors?", multiple: true, required: false
+        input "locks", "capability.lock", title: "Which Locks?", multiple: true, required: false
+    	input "garageDoors", "capability.garageDoorControl", title: "Which Garage Doors?", multiple: true, required: false
+		input "cameras", "capability.videoCamera", title: "Which Cameras?",  multiple: true, required: false
+    }
+}
+
+//API external Endpoints
+mappings {
+	path("/subscriptionURL/:url") {
+        action: [
+            PUT: "updateEndpointURL"
+        ]
+    }
+    path("/connectionId/:connId") {
+        action: [
+            PUT: "updateConnectionId"
+        ]
+    }
+	path("/devices") {
+		action: [
+			GET: "getDevices"
+		]
+	}
+    path("/devices/:id") {
+		action: [
+			GET: "getDevice"
+		]
+	}
+    path("/update/:id") {
+        action: [
+          PUT: "updateDevice"
+        ]
+    }
+}
+
+def installed() {
+	log.debug "Installed with settings: ${settings}"
+	initialize()
+}
+
+def updated() {
```

Comment:
Note. this method is called every time user opens the SmartApp and then click "Done" in upper right corner to close it. Hence, initialized() will also be called and  state.connectionId/endpoint will be set to default values. Just highlighting this since I do not know when updateEndpointURL/updateConnectionId are called.